Integration test for masternode operations

domob1812 opened the pull request:
This adds a new integration test for basic operation of masternodes:  Funding, starting, syncing and payments.

To make masternode operations work properly on regtest, we had to loosen some restrictions (e.g. allow same host / different IP nodes in netfulfilledman) for nodes running on regtest.

domob1812 committed:
Minor cleanups in masternode code.

Some small cleanups in the masternode code that do not change code
behaviour but make it easier to read.

domob1812 committed:
Allow masternodes on arbitrary ports.

There is not really any use in restricting masternodes to the
default P2P port, and it hurts regtest-ability.

domob1812 committed:
New test for masternode operation.

The new mnoperation.py integration test funds and starts masternodes,
and verifies that they get added to the masternode list successfully,
masternode sync goes through, and the payments match expectations.

domob1812 committed:
Use GetTime() for mn sync process.

Instead of using ticks in CMasternodeSync::Process and assuming the
function is called every second, use GetTime() compared to a known
timestamp of the last sync step to trigger the individual stages.

This is much clearer to understand, more explicit, and it also enables
using mocktime more easily for regtests.

Also includes some general code cleanups:  Two constants that are used
only internally in masternode-sync.cpp are moved to the cpp file to
make this explicit, and since we have a class, use a member variable
instead of a function-static one to hold the last call timestamp.

domob1812 committed:
Use GetTime() in masternode handler thread.

Instead of counting 1-second-sleeps to time operations, use
GetTime() and recorded timestamps of when something was last
done in the masternode handler thread (ThreadCheckObfuScationPool).
This makes the code clearer to read and the intent more explicit.

It also makes it easier to use mocktime for testing.  To further
facilitate that, we even use a condition variable to wake the sleeping
thread immediately if the mocktime is changed.

domob1812 committed:
Remove regtest special case on masternode sync.

We do not want to special case regtest for masternode sync, since
that prevents us from actually testing the real process.

domob1812 committed:
Allow same IP in netfulfilledman on regtest.

The CNetFulfilledRequestManager keeps track of which peers have already
gotten / sent which requests, to avoid DoS vectors.  It tracks peers
by IP only, not the full IP/port address.

On regtest, this is obviously insufficient, as many peers will by design
have the same IP address (localhost).  Thus on regtest, we want to track
by the full network address.

domob1812 committed:
Disable masternode scheduled check on regtest.

Usually, masternodes are only considered as payment winner if they
are not already receiving another payout in the next eight blocks.
On regtest, this is annoying -- because when testing with a very small
number of masternodes, it means that those will always be scheduled and
also skipped because of that check.  (And worse, the rule can even force
particular non-random patterns in the payouts.)

This change disables the scheduled check on regtest, so it is easier
to test things like the payout schedule / chances.

domob1812 committed:
Reduce sleep when syncing in regtests.

When syncing blocks and/or the mempool between nodes in the regtest
framework, we repeatedly query to see if they match and sleep if not.

This reduces the sleep from 1s to 100ms, which makes in general makes
syncing faster (but the sleep is still more than long enough to prevent
a "busy wait" loop).

galpHub commented:
Looks like an issue with a missing include
libbitcoin_wallet.a(libbitcoin_wallet_a-obfuscation.o): In function `ThreadCheckObfuScationPool()':
/usr/include/boost/thread/pthread/condition_variable_fwd.hpp:251: undefined reference to `boost::chrono::system_clock::now()'

domob1812 committed:
Fix boost chrono linking.

With the usage of boost::condition_variable (from boost thread), we now
need to always link in boost chrono on which it depends.  Also the
ordering of libraries was wrong for test_divi, we should place boost
and the other dependencies after the wallet library.
